lib/Interpreter/IncrementalCUDADeviceCompiler.cpp

```diff
@@ -0,0 +1,449 @@
+    llvm::raw_fd_ostream dummyCU(m_DummyCUPath, EC, llvm::sys::fs::F_Text);
+    if(EC){
+      llvm::errs() << "Could not open " << m_DummyCUPath << ": "
+        << EC.message() << "\n";
+      return false;
+    }
+    dummyCU.close();
+
+    m_PTXFilePath = m_FilePath + "cling.ptx";
+    m_GenericFileName = m_FilePath + "cling";
+    return true;
+  }
+
+  bool IncrementalCUDADeviceCompiler::searchCompilingTools(
+      cling::InvocationOptions & invocationOptions){
+    // Search after clang in the folder of cling.
+    llvm::SmallString<128> cwd;
+    // get folder of the cling executable to find the clang which is contained
+    // in cling
+    // nullptr is ok, if we are the main and not a shared library
+    cwd.append(llvm::sys::path::parent_path(
+      llvm::sys::fs::getMainExecutable(
+        invocationOptions.CompilerOpts.Remaining[0], (void *) &cwd)));
+    cwd.append(llvm::sys::path::get_separator());
+    cwd.append("clang++");
+    m_ClangPath = cwd.c_str();
+    // Check, if clang is existing and executable.
+    if(!llvm::sys::fs::can_execute(m_ClangPath)){
+      llvm::errs() << "Error: " << m_ClangPath << " not existing or executable!\n";
+      return false;
+    }
+
+    // Use the custom CUDA toolkit path, if it set via cling argument.
+    if(!invocationOptions.CompilerOpts.CUDAPath.empty()){
+      m_FatbinaryPath = invocationOptions.CompilerOpts.CUDAPath + "/bin/fatbinary";
+      if(!llvm::sys::fs::can_execute(m_FatbinaryPath)){
+        llvm::errs() << "Error: " << m_FatbinaryPath << " not existing or executable!\n";
+        return false;
+      }
+    }else{
+      // Search after fatbinary on the system.
+      if (llvm::ErrorOr<std::string> fatbinary =
+            llvm::sys::findProgramByName("fatbinary")) {
+        llvm::SmallString<256> fatbinaryAbsolutePath;
+        llvm::sys::fs::real_path(*fatbinary, fatbinaryAbsolutePath);
+        m_FatbinaryPath = fatbinaryAbsolutePath.c_str();
+      } else {
+        llvm::errs() << "Error: nvidia tool fatbinary not found!\n" <<
+          "Please add the cuda /bin path to PATH or set the toolkit path via --cuda-path argument.\n";
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  void IncrementalCUDADeviceCompiler::addHeaders(
+      llvm::SmallVectorImpl<std::string> & argv){
+    for(clang::HeaderSearchOptions::Entry e : m_HeaderSearchOptions->UserEntries){
+      if(e.Group == clang::frontend::IncludeDirGroup::Angled)
+        argv.push_back("-I" + e.Path);
+    }
+  }
+
+  bool IncrementalCUDADeviceCompiler::generateFatbinary(const llvm::StringRef input,
+                                                        cling::Transaction * T){
+    if(!m_Init){
+      llvm::errs() << "Error: Initializiation of CUDA Device Code Compiler failed\n";
+      return false;
+    }
+
+    // Write the (CUDA) C++ source code to a file.
+    std::error_code EC;
+    llvm::raw_fd_ostream cuFile(m_GenericFileName + std::to_string(m_Counter)
```

Comment:
`getAndIncrCounter` is not a good idea. I just increase `m_counter`, if `generateFatbinray` has no error, because I assume every .cu-file from cling0.cu to cling<m_counter>.cu is correct. So, in every failure case, I have to decrease the variable and thats sounds really error-prone. 

Instead I would use a `const unsigned int IncrementalCUDADeviceCompiler::getImmutableCounter()` function to get a copy of the counter and only use the `m_counter` variable direct, if I increase it.